src/node/historical_queries.h

```diff
@@ -0,0 +1,341 @@
+      {
+        throw std::logic_error(fmt::format(
+          "Signature {} claims it was produced by node {}: This node is "
+          "unknown",
+          sig_idx,
+          sig->node));
+      }
+
+      auto verifier = tls::make_verifier(node_info->cert);
+      const auto verified = verifier->verify_hash(
+        real_root.h.data(),
+        real_root.h.size(),
+        sig->sig.data(),
+        sig->sig.size());
+      if (!verified)
+      {
+        throw std::logic_error(
+          fmt::format("Signature at {} is invalid", sig_idx));
+      }
+
+      // TODO: Find which of our untrusted indices are in this tree
+      auto it = requests.begin();
+      while (it != requests.end())
+      {
+        auto& request = it->second;
+
+        if (request.current_stage == RequestStage::Untrusted)
+        {
+          const auto& untrusted_idx = it->first;
+          const auto& untrusted_hash = request.entry_hash;
+          const auto& untrusted_store = request.store;
+
+          try
+          {
+            const auto receipt = tree.get_receipt(untrusted_idx);
+            LOG_INFO_FMT(
+              "From signature at {}, constructed a receipt for {}",
+              sig_idx,
+              untrusted_idx);
+
+            // TODO: Check that the receipt matches our untrusted entry
+            // const auto& entry_hash_in_receipt = receipt.leaf;
+            // if (untrusted_hash != entry_hash_in_receipt)
+            // {
+            //   throw std::logic_error(
+            //     fmt::format("Hash mismatch for {}", untrusted_idx));
+            // }
+
+            // Move stores from untrusted to trusted
+            // TODO: Temp solution, blindly trust everything for now
+            LOG_INFO_FMT("Now trusting {}", untrusted_idx);
+            request.current_stage = RequestStage::Trusted;
+            ++it;
+          }
+          catch (const std::exception& e)
+          {
+            LOG_INFO_FMT(
+              "Signature at {} does not cover {}: {}",
+              sig_idx,
+              untrusted_idx,
+              e.what());
+            // TODO: Can we expose "what indices do you cover" through the
+            // MerkleTree, rather than try-catch?
+            ++it;
+            // TODO: Should we abandon this untrusted entry now?
+          }
+        }
+        else
+        {
+          ++it;
+        }
+      }
+    }
+
+    void deserialise_ledger_entry(
+      consensus::Index idx, const LedgerEntry& entry)
+    {
+      StorePtr store = std::make_shared<kv::Store>();
+
+      store->set_encryptor(source_store.get_encryptor());
+
+      // TODO: Add a lazy clone option?
+      store->clone_schema(source_store);
+      store->set_strict_versions(false);
+
+      const auto deserialise_result = store->deserialise_views(entry);
+
+      switch (deserialise_result)
+      {
+        case kv::DeserialiseSuccess::FAILED:
+        {
+          // TODO: Host gave us junk? Do we fail silently?
```

Comment:
I vote against this. 
This seems to be on the same level of crashing if we get a malformed message

Reply:
We can't avoid throwing if the host gives us malformed data - `deserialise_views` will throw on many kinds of malformed data, the merkle tree can only report failures by throwing, so its no more dangerous to throw here as well. But I agree this shouldn't propagate up to crashing the node, so I've added a catch for these failures at the top-level `handle_ledger_entry`.